server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java

```diff
@@ -85,14 +85,16 @@ protected void acquireReplicaOperationPermit(final IndexShard replica,
     }
 
     @Override
-    protected PrimaryResult<ShardRequest, ReplicationResponse> shardOperationOnPrimary(final ShardRequest shardRequest,
-                                                                                       final IndexShard primary) throws Exception {
-        executeShardOperation(shardRequest, primary);
-        return new PrimaryResult<>(shardRequest, new ReplicationResponse());
+    protected void shardOperationOnPrimary(final ShardRequest shardRequest, final IndexShard primary,
+            ActionListener<PrimaryResult<ShardRequest, ReplicationResponse>> listener) {
+        ActionListener.completeWith(listener, () -> {
```

Comment:
Since most of write actions execute its primary action synchronously, maybe introduce an async method which uses the result from the sync version by default (see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java#L106-L120).

Reply:
I tried doing this here, but ran into a bit of a mess with the slightly strange way we handle the overrides of this method where we constrain the type to `WritePrimaryResult` so I decided to not do it here to unblock another thing.
I like the idea though! => I'll open a PR that cleans this up (very) shortly :)

Reply:
No worry. It's optional. Thanks for giving it a try.